### PR TITLE
Added xgb-training code for predicting cve severity

### PR DESCRIPTION
**Pull request by suppathak**
This PR adds an XGBoost classifier for predicting Linux kernel CVE severity levels (IMPORTANT/MODERATE/LOW).

## What it does:
- Trains XGBoost model using 48 binary features extracted from CVE patch analysis
- Features include file path analysis, security indicators, and code patterns
- Outputs trained model, metadata, and feature importance rankings

## Usage:
```bash
uv run python src/xgb-classifier/xgboost-train.py

### CODE REVIEW

File: src/xgb-classifier/README.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+- **MODERATE (1)**: Medium severity vulnerabilities  
+- **LOW (2)**: Low severity vulnerabilities
+
+## Data Pipeline & Feature Engineering
+
+The training data is constructed through a multi-step process:
+
+1. **CVE Data Collection**: Scrapes Linux kernel CVE metadata from security repositories
+2. **Commit & Patch Retrieval**: Fetches actual commit patches and details from the Linux kernel Git repository
+3. **Feature Engineering**: Analyzes patch content and commit messages to extract 48 binary features:
```
Comment (erikerlandson) [inline src/xgb-classifier/README.md:18]: minor nit: this is "feature extraction"  
"feature engineering" is the process of creating or curating features to use

**Comment by hemajv** [comment]
@JimFuller-RedHat this PR only includes the code for training the XGBoost model
let us know if you would like us to include the prediction code as well or create a separate PR for it

File: src/xgb-classifier/xgboost-train.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        logger.error(f"Training failed: {e}")
+        return
+    
+    # Create models directory
+    models_dir = Path('models')
+    models_dir.mkdir(exist_ok=True)
+    logger.info(f"Created models directory: {models_dir.absolute()}")
+    
+    # Save the model
+    model_path = models_dir / 'cve_severity_model.pkl'
```
Comment (JimFuller-RedHat) [inline src/xgb-classifier/xgboost-train.py:89]: it would be good to push any saved models to a `src/aegis_ai_ml/etc/models` directory

**Comment by JimFuller-RedHat** [comment]
> @JimFuller-RedHat this PR only includes the code for training the XGBoost model let us know if we should include the prediction code as well or create a separate PR for it

please include with this PR the entire thing ... we are happy to take what you have and extrude general functions into tooling

File: src/xgb-classifier/README.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
+# CVE Severity XGBoost Classifier
+
+XGBoost classifier for predicting Linux kernel CVE severity levels based on patch analysis and feature extraction.
+
+## Overview
+
+This classifier predicts CVE severity into three categories:
```
[the remark this answers is not in the document]
Comment (erikerlandson) [reply src/xgb-classifier/README.md:7]: yes, there are no critical examples in any of the data

File: pyproject.toml [hunk 4]
```diff
@@ -29,6 +29,9 @@ dependencies = [
     "wikipedia>=1.4.0",
     "logfire[fastapi,httpx,sqlite3]>=4.0.0",
     "platformdirs>=4.3.8",
+    "pandas>=1.5.0",
```
[the remark this answers is not in the document]
Comment (suppathak) [reply pyproject.toml:32]: will do so! Thank you.

**Comment by suppathak** [comment]
> please include with this PR the entire thing ... we are happy to take what you have and extrude general functions into tooling

Sure will do!! Thank you @JimFuller-RedHat .

File: src/xgb-classifier/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (JimFuller-RedHat) [reply src/xgb-classifier/README.md:7]: yes, just wanted to check this is baked in (which is fine!)

**Review by erikerlandson — COMMENTED** [review]
Maybe we should rename the directory `xgb-classifier` to something a bit more specific, for example `cve-impact-classifier`.  That specifies what it's for (and ignores the implementation detail like "xgb")

**Comment by suppathak** [comment]
> Maybe we should rename the directory `xgb-classifier` to something a bit more specific, for example `cve-impact-classifier`. That specifies what it's for (and ignores the implementation detail like "xgb")

Thank you @erikerlandson . Make sense. i made the required change.

File: src/aegis_ai_ml/src/classifier/cve-impact-classifier/cve_data_scraper.py [hunk 5]
```diff
@@ -0,0 +1,673 @@
...
+        # Look for commit references in various fields
+        def extract_commits_from_text(text):
+            if not text:
+                return []
+            
+            commit_patterns = [
+                r'https://git\.kernel\.org/[^/]+/[^/]+/[^/]+/[^/]+/[^/]+/commit/\?id=([0-9a-fA-F]+)',
+                r'https://git\.kernel\.org/stable/c/([0-9a-fA-F]+)',
+                r'https://github\.com/torvalds/linux/commit/([0-9a-fA-F]+)',
+                r'commit\s+([0-9a-fA-F]{40})',
```
Comment (kdudka) [inline src/aegis_ai_ml/src/classifier/cve-impact-classifier/cve_data_scraper.py:246]: This seems wrong.  We do not want to extract arbitrary commit IDs from the text.  We want to examine the commit(s) that **fixed** the CVE.  For example with [CVE-2024-58093](https://lore.kernel.org/linux-cve-announce/2025041652-CVE-2024-58093-2638@gregkh/T/#u), this extracts (also) all commit IDs from the following text:
```
Affected and fixed versions
===========================

	Issue introduced in 6.5 with commit 456d8aa37d0f56fc9e985e812496e861dcd6f2f2 and fixed in 6.15-rc1 with commit cbf937dcadfd571a434f8074d057b32cd14fbea5
	Issue introduced in 5.4.251 with commit 666e7f9d60cee23077ea3e6331f6f8a19f7ea03f
	Issue introduced in 5.10.188 with commit 7badf4d6f49a358a01ab072bbff88d3ee886c33b
	Issue introduced in 5.15.121 with commit 9856c0de49052174ab474113f4ba40c02aaee086
	Issue introduced in 6.1.39 with commit 7aecdd47910c51707696e8b0e045b9f88bd4230f
	Issue introduced in 6.3.13 with commit d51d2eeae4ce54d542909c4d9d07bf371a78592c
	Issue introduced in 6.4.4 with commit 4203722d51afe3d239e03f15cc73efdf023a7103
```

This includes the commits that **introduced** the CVE.  I think it is unlikely that they would help with estimation of the impact.  In vast majority of cases, the authors were not aware of the fact they were introducing a CVE when they were creating these commits.

Comment (suppathak) [reply src/aegis_ai_ml/src/classifier/cve-impact-classifier/cve_data_scraper.py:246]: You are right @kdudka . The current code extracts all commits including those that introduced the CVE. I will update it to specifically look for fixing commits.

**Review by hemajv — COMMENTED** [review]
@suppathak I think it would be good to also include a `README` for the `data` folder to provide more clarity around their use/purpose

similar to @erikerlandson comment above, since we are doing the severity analysis for only kernel CVEs, we should update the dir to `kernel-cve-impact-classifier` instead

File: src/aegis_ai_ml/src/classifier/cve-impact-classifier/cve_data_scraper.py (hunk 5 above)
Comment (kdudka) [reply src/aegis_ai_ml/src/classifier/cve-impact-classifier/cve_data_scraper.py:246]: I was thinking mainly about the commit messages.  It is usually the commit message of the CVE fix what describes the cause, consequence, fix, and result, which is IMO the reliable source of information to assess impact of the CVE.  The commit message associated with the change that introduced the CVE usually says "implement such and such enhancement", "fix such and such bug", "migrate the code to a new version of some interface", etc.  As embarrassing as it is, there are also cases where a CVE fix introduces another CVE of a different impact level.  For example, we successfully fix a resource leak but, at the same time, we introduce a use after free on some error handling path.

Looking at `cve_feature_extraction.py` now, I can see that the script looks at the actual code changes, too.  This might be useful for the `introduced in ...` commits in case someone introduces a CVE by doing small changes in existing code.  It will not help when a big piece of fresh code is added such that it includes the CVE from the beginning.  In any case, we should not handle the properties extracted from both kinds of commits in the same way.

File: src/aegis_ai_ml/src/classifier/cve-impact-classifier/cve_feature_extraction.py [hunk 6]
```diff
@@ -0,0 +1,758 @@
...
+                labels[cve['cve_id']] = cve['impact']
+            print(f"Loaded {len(train_data)} CVEs with labels")
+        except FileNotFoundError:
+            print("kernel_cves.json not found")
+
+        return labels
+    def clean_content(self, content: str) -> str:
+        """Clean and normalize content"""
+        # limiting size for mem 
+        if len(content) > 10 * 1024 * 1024:  # 10MB limit
```
Comment (kdudka) [inline src/aegis_ai_ml/src/classifier/cve-impact-classifier/cve_feature_extraction.py:72]: This duplicates the size check in `extract_patch_features()` but with higher threshold.  So the `if` body is unreachable.

**Comment by suppathak** [comment]
> The except branch is unreachable unless subprocess.run() is called with check=True. Overall I do not understand the code flow here. The code is not fetching a specific refspec. If the first command fails to fetch 50 commits below HEAD in 30s, what is the point in trying to fetch an unlimited number of commits in 60s? How does any of those commands ensure that the requested commit will become available?

> This duplicates the size check in extract_patch_features() but with higher threshold. So the if body is unreachable.

@kdudka , I have address these two feedback in this commit.  ptal. Thank you.

**Comment by suppathak** [comment]
@hemajv , I have updated the folder name and added a readme for data folder as well. ptal! Thank you!

**Review by hemajv — APPROVED** [review]
@suppathak thanks for adding a README to explain the `data` dir
/lgtm 👍

File: src/aegis_ai_ml/src/classifier/kernel-cve-impact-classifier/README.md [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+- Extracts CVE details, commit URLs, and patch information
+- Saves structured data with patches and metadata for each CVE
+
+### 2. Feature Extraction ([`cve_feature_extraction.py`](cve_feature_extraction.py))
+- Processes scraped CVE data to extract 48 binary features:
+ - **File Path Analysis**: Component types (networking, hardware, disk, etc.)
+ - **Patch Content Analysis**: Security indicators (kasan, uaf, nullptr, danger, etc.)
+ - **Code Pattern Analysis**: Code changes (lock, memory, timer, init patterns)
+ - **Metadata Features**: CVE timing, author info, fix detection patterns
+- Outputs: [`cve-training-dataset.csv`](data/cve-training-dataset.csv) and [`cve_testing_dataset.csv`](data/cve_testing_dataset.csv)
```
[the remark this answers is not in the document]
Comment (suppathak) [reply src/aegis_ai_ml/src/classifier/kernel-cve-impact-classifier/README.md]: @kdudka, addressed the filename consistency issue! PTAL.
Thank you!

**Comment by suppathak** [comment]
@kdudka Just a quick reminder, please lmk if there are any additional changes needed. If everything looks good, would you be able to approve and merge?
Thank you.

**Comment by JimFuller-RedHat** [comment]
@suppathak - would you be able to correct the test and ruff lint errors ? They look mostly minor like unused imports.

**Comment by suppathak** [comment]
> To be honest, I have not yet been able to reproduce the experiment end-to-end. Let me open a separate pull request to address the outstanding issues and to compare my results with the expected results.

Thanks @kdudka. I’m planning to record a quick end-to-end demo by running all the scripts sequentially and will share the recording once it’s ready. If I come across any gaps during the process, I’ll address them in the next PR.
